File: php/5.8-convert_base/convert_base.php
<?php

require_once '../bootstrap.php';

use EOPI\Helper\InputHelper;

/**
 * The time complexity is O(n(1 + log(base base2)base1)), where n is the length of s.  The reasoning is as follows.
First, we perform n multiply-and-adds to get x from s.  Then we perform log(base base2)x multiply and adds to get
the result.  The value x is upper-bounded by base1^n, and log(base base 2)base1^n = nlog(base base2)base1
 */

/**
 * Convert a number from one base to another.  This works by first converting the number to a decimal (base 10).  Once
 * the number is in decimal, it is converted to the new base.
 * To convert to decimal multiply each digit of the original number, multiply each value by the base raised to k.  For
 * example 173 in base 5 becomes 3 * 5^0 + 7 * 5^1 + 1 * 5^2 or 3 * 1 + 7 * 5 + 1 * 25 or 3 + 35 + 25 or 63.
 * To convert from decimal to the new base, go divide the decimal by the new base and find the remainder then add that
 * to the new number.  Do this until the decimal is <= 0.  Then reverse the new number because the values were added in
 * reverse order.  For example if the decimal is 63 and the new base is 8:
 * 63 % 8 is 7 and floor(63 / 8) = 7
 * 7 % 8 is 7 and 7 / 8 is less than 0
 * The new number is 77
 * To handle values larger than 9 this uses the translation array.
 *
 * @param string $number
 * @param int $base1
 * @param int $base2
 * @return string
 * @throws \InvalidArgumentException
 * @throws \OutOfRangeException
 */
function convertBase($number, $base1, $base2)
{
    if (!is_int($base1)) {
        throw new \InvalidArgumentException('$base1 must be an integer');
    }

    if ($base1 < 2 || $base1 > 16) {
        throw new \OutOfRangeException('$base1 must be between 2 and 16 inclusive');
    }

    if (!is_int($base2)) {
        throw new \InvalidArgumentException('$base2 must be an integer');
    }

    if ($base2 < 2 || $base2 > 16) {
        throw new \OutOfRangeException('$base2 must be between 2 and 16 inclusive');
    }

    $isNegative = $number[0] == '-';
    $characterTranslations = [
        'A' => 10,
        'B' => 11,
        'C' => 12,
        'D' => 13,
        'E' => 14,
        'F' => 15,
    ];
    $valueTranslations = array_flip($characterTranslations);

    // Convert the number to decimal (base 10)
    $decimal = 0;

    for ($i = ($isNegative ? 1 : 0); $i < strlen($number); $i++) {
        $decimal *= $base1;
        $decimal += is_numeric($number[$i]) ? (int)$number[$i] : $characterTranslations[$number[$i]];
    }

    if (!is_integer($decimal)) {
        throw new \OutOfRangeException('converting to decimal produced a number that is too large');
    }

    // Convert the decimal to the new base
    $newNumber = '';

    while ($decimal) {
        $remainder = $decimal % $base2;
        $newNumber .= $remainder < 10 ? $remainder : $valueTranslations[$remainder];
        $decimal = (int)($decimal / $base2);
    }

    if ($isNegative) {
        $newNumber .= '-';
    }

    return strrev($newNumber);
}

$inputHelper = new InputHelper();
$number = $inputHelper->readInputFromStdIn('Enter the string in base 1: ');
$base1 = $inputHelper->readInputFromStdIn('Enter the integer base 1 between 2 and 16 inclusive: ');
$base2 = $inputHelper->readInputFromStdIn('Enter the integer base 2 between 2 and 16 inclusive: ');
$result = convertBase((string)trim($number), (int)$base1, (int)$base2);
printf('The result of converting %s from base %d to %d is: %s', $number, $base1, $base2, $result);
print PHP_EOL;

File: python2/5.8-convert_base/convert_base.py
"""
Convert a number from one base to another where both bases are between 2 and 16 (inclusive).

The time complexity is O(n(1 + log(base base2)base1)), where n is the length of s.  The reasoning is as follows.
First, we perform n multiply-and-adds to get x from s.  Then we perform log(base base2)x multiply and adds to get
the result.  The value x is upper-bounded by base1^n, and log(base base 2)base1^n = nlog(base base2)base1
"""

__author__ = "Matt Rathbun"
__email__ = "mrathbun80@gmail.com"
__version__ = "1.0"


def convert_base(number, base1, base2):
    """
    Convert number from base1 to base2
    i.e. 245 in base 6 to base 14 is:

    first convert the number to decimal (base 10)
    -------------------------------------------------
    = 2 * 6^2 + 4 * 6^1 + 5 * 6^0
    = 2 * 36 + 4 * 6 + 5 * 1
    = 72 + 24 + 5
    = 101 in base 10

    then convert the decimal to base 14
    -------------------------------------------------
    = 101 / 14 = 7 R 3
    = 73 in base 14

    i.e. 1045 in base 8 to base 9 is:

    first convert the number to decimal (base 10)
    -------------------------------------------------
    = 1 * 8^3 + 0 * 8^2 + 4 * 8^1 + 5 * 8^0
    = 1 * 512 + 0 * 64 + 4 * 8 + 5 * 1
    = 512 + 0 + 32 + 5
    = 549 in base 10

    then convert the decimal to base 9
    -------------------------------------------------
    = 549 / 9 = 61 R 0
    = 61 / 9 = 6 R 7
    = 670 in base 9

    HOW IT WORKS
    -------------------------------------------------
    Converting from the original base to decimal happens left to right by performing the multiplication by the base
    first in each iteration.  Using the first example from above (245 in base 6 to base 14) we have:

    Iteration 1:
    x *= 6 / x = 0
    x += 2 / x = 2

    Iteration 2:
    x *= 6 / x = 12
    x += 4 / x = 16

    Iteration 3:
    x *= 6 / x = 96
    x += 5 / x = 101

    Doing this the 2 gets multiplied by the base twice or 2 * 6^2, 4 gets multiplied by the base once or 4 * 6^1 and 5
    gets multiplied by the base never or 5 * 6^0.  Yielding 2 * 6^2 + 4 * 6^1 + 5 * 6^0 which is the same as the
    elementary formula.

    -------

    Coverting from the decimal to the new base happens by repeatedly computing the remainder at each position.  Using
    the first example from above (245 in base 6 to base 14.  Decimal value 101) we have:

    Iteration 1:
    remainder = 101 % 14 / remainder = 3 or (98 R 3)
    result = 3
    x = 101 / 14 / x = 7

    Iteration 2:
    remainder = 7 % 14 / remainder = 7 or (0 R 7)
    result = 37
    x = 7 / 14 / x = 0

    The result is then reversed for a value of 73
    """
    is_negative = number[0] == '-'
    x = 0
    start = 0
    character_translations = {
        'A': 10,
        'B': 11,
        'C': 12,
        'D': 13,
        'E': 14,
        'F': 15
    }
    value_translations = {
        10: 'A',
        11: 'B',
        12: 'C',
        13: 'D',
        14: 'E',
        15: 'F'
    }

    if is_negative:
        start = 1

    # Convert the number to a decimal
    for i in range(start, len(number)):
        x *= base1

        try:
            x += int(number[i])
        except ValueError:
            x += character_translations[number[i].upper()]

    result = []

    # Convert the decimal to the new base
    while x:
        remainder = x % base2

        if remainder >= 10:
            remainder = value_translations[remainder]

        result.append(remainder)
        x /= base2

    if is_negative:
        result.append('-')

    result.reverse()

    return ''.join(str(value) for value in result)


number = str(raw_input('Number: '))
base1 = int(raw_input('Base 1: '))
base2 = int(raw_input('Base 2: '))
print 'The number %s in base %d is %s in base %d' % (number, base1, convert_base(number, base1, base2), base2)
